### ISSUE TITLE
Cmdsets with mergetype 'remove' do not work on caller

### ISSUE THREAD

**Issue by Nadorrano** [evennia/evennia#1087]
#### Brief summary of issue:
The problem is that when an object has a cmdset with the "remove" flag on the top of the stack (highest priority), the flag is propagated to obj.cmdset.current, causing unexpected behaviors when obj is caller.
This is what happen if I put a "remove" cmdset on a character:
 - cmdsets are collected from local objects and merged on each object, this is OK
 - cmdsets stored on caller are merged, the "remove" cmdset is merged last and the flag is retained by the merged cmdset
 - cmdsets on caller and objects are merged together, but in the previous step caller has got a cmdset containing all his commands and the flag "remove",
 in this merge it will behave as a filter not as regular cmdset!

This has the effect of removing basically all commands on caller.
The same happens if the cmdset is stored on external objects with same priority as the caller current cmdset, because it fails when merging cmdsets with same priority (this may be a separate issue).

#### Steps to reproduce the issue / Reasons for adding feature:
This command will add a CmdSet with a high priority (must be the highest of all cmdsets stored on 'character') to a character. 
The result is that the character lose its commands.

`@py from evennia import CmdSet; a = CmdSet(); a.priority = 40; a.mergetype = "Remove"; self.search("char name").cmdset.add(a)`

The command is a single line.

#### Extra information, such as Evennia revision/repo/branch, operating system and ideas for how to solve / implement:

As a quickfix I found that forcing the caller (and only the caller) to have caller.cmdset.current.mergetype = "Union", solves this issue.

I hope this bug description is clear.

**Comment by Nadorrano** [comment]
Hi again,
before reporting this bug I wanted to file a feature request. I've realized just now that the feature would probably solve this issue. I'll explain briefly.
The problem is that currently cmsets from local objects does not "interpose" with caller's cmdsets.
As an example image the dark room in the tutorial, that adds a cmdset "replace" on characters in the room, and let's say this cmdset has priority 10. Let's also say that our character has the default cmdset, priority 0, and another cmdset containing a super important command, with priority 20. Well, in this case the character's merged cmdset would result having priority 20 and the dark room cmdset does not kick in.

Wouldn't be better if cmdsets on the caller (and only the caller) were mixed with local objects cmdsets, separately? (I don't mean to merge every cmdset separately, just to mix already-merged cmdsets of local objects with cmdsets on caller and then merge all together.)

I think that this will solve the bug because it makes the merging order more linear, but I have not tried it and I don't know if this will break something.

**Comment by Griatch** [comment]
@Nadorrano I have been pondering this a bit in the last week and your suggestion of merging all command sets in one "batch", so to speak, does make sense from a usability standpoint. There is one possible snag though: The original reason for merging object cmdsets separately is that those are actually treated a little differently; they are forcibly set to allow duplicates. The thinking goes that if you have one red and and one blue button in the room, both with a same-prio "press" command on them, you *don't* want to treat it as the same command and merge them. If you did the button you press would just be due to which of the two happened to be merged last in the queue (which is unknown to the user). So instead the system allows both such commands to co-exist so that when you use the command "press" you must specify if you want the command related to the red or blue button. 

I agree though that your suggestion would likely solve the issue presented here. I've not looked at it in detail, but it's possible one can merge in the extra cmdsets (channels, exits, objects) in a similarly "homogenous" way while preserving what makes them special. I'd need to look into that further.

**Comment by Griatch** [comment]
@Nadorrano 

I have experimented with this a bit in a local branch: I've reworked how cmdset options and mergetypes traverse the stack and reworked cmdhandler.get_and_merge_cmdsets to build a long list from all cmdsets (regardless of source) instead of pre-merging them on each object and then adding them. The .current cmdsets on the caller still has a role since the final on-caller meged set determines the status of the various options (`no_exits`, `no_channels`, `no_objs`), options which traverse freely up the stack from a low-prio cmdset unless a higher-prio set explicitly changes them. 

There are some quirks to work out but it looks promising. It also got me inspired for writing unit tests for the command system which is a good thing in itself. I might put the result up in a devel branch for people to try out, in case the subtle changes herein affect people in some unexpected ways. Commands and Command set merging is after all a pretty core thing.
